### DISCUSSION TITLE
Support partial indexes in onConflict clauses

### DISCUSSION THREAD

**Discussion by chrisbenincasa** [mikro-orm/mikro-orm#5668] [Ideas]
**Is your feature request related to a problem? Please describe.**
Partial indexes in SQLite and PSQL require specifying the `where` clause originally used when creating the index. There is currently no way to express these type of `ON CONFLICT` cause in mikro-orm. There is [knex support](https://knexjs.org/guide/query-builder.html#onconflict) using "raw".

**Describe the solution you'd like**
One possible solution would be to add a new `onConflictWhere` option (or some other name) to the `upsert` options. This option would only be valid if the `onConflictFields` option was also specified. `onConflictWhere` could simply take a raw SQL expression string that gets plumbed into the Knex query. 

**Describe alternatives you've considered**
Partial indexes don't really have first-class support in the framework as it is and Knex only supports using them in the query builder using raw strings, so I'm not sure what other options there are that wouldn't require a substantial amount of work / refactoring (changing how index declaration works on entities, which I've seen elsewhere is not really desired work for mikro-orm)

**EDIT**: It looks like perhaps this could also just be supported in the query builder, since it looks like the onConflict param type [already has a field for a where clause](https://github.com/mikro-orm/mikro-orm/blob/42091b13d2a33ce5e47d802af42c95b15a77edea/packages/knex/src/query/QueryBuilder.ts#L182). Perhaps this can just be exposed in the API and then plumbed down?

**Additional context**
N/A

**Comment by B4nan** [comment]
Can you provide a complete example of what you want, e.g. by using knex? 

The `where` in `onConflict` refers to `qb.where()` condition:

https://github.com/mikro-orm/mikro-orm/blob/42091b13d2a33ce5e47d802af42c95b15a77edea/packages/knex/src/query/QueryBuilder.ts#L513

Here is a test case that uses it:

https://github.com/mikro-orm/mikro-orm/blob/42091b13d2a33ce5e47d802af42c95b15a77edea/tests/QueryBuilder.test.ts#L2758-L2770

**Comment by chrisbenincasa** [comment]
There is a Knex example in the knew docs I linked for this exact situation. 

```
knex('tableName')
  .insert({
    email: 'ignore@example.com',
    name: 'John Doe',
    active: true,
  })
  // ignore only on email conflict and active is true.
  .onConflict(knex.raw('(email) where active'))
  .ignore();
```

<details>
<summary>Knex docs blurb about partial indexes</summary>
<img src="https://github.com/mikro-orm/mikro-orm/assets/1640671/609be02d-5da9-48e7-ac70-82b21d0ec56f" />
</details>

**Reply by B4nan** [reply]
Yes, and I am asking for a complete example I can run myself and experiment with, example based on your needs, not example from knex docs. A screenshot is not a complete reproduction, neither is a link to docs (not to mention docs of other tools).

I gave you an example of what can be done now, have you checked whether it fits your needs? Because it feels like its doing the same thing to me.

The fact that this is supported in knex only via `raw` is not very convincing that we need more than that. I can imagine you could pass instance of `knex.raw` the same way in the end.

**Reply by chrisbenincasa** [reply]
OK, I did not see your edit and also was not aware that a full reproduction was needed (I just followed the template for the feature request). Unfortunately, I don't think the `where` there will do exactly what I'm looking for here. The generated query would need to have the `where` clause come right after the on conflict fields in order to properly match against the definition of the unique index

Here is a lighter example of the situation I'm faced with: https://github.com/chrisbenincasa/reproduction-populate-hint/pull/new/partial-index

I described a bit of the situation in a comment at the top of the test file, but the TL;DR is that NULLs in unique keys in SQLite are not considered in 'uniqueness'. The solution for this that I was working on was two partial indexes, one when for when a column is NULL and when for NOT NULL. It's not the most elegant, since it'll require a bit more tedious app code (splitting generated entities by which ones have NULL for the given column in order to choose to the correct partial index in the ON CONFLICT clause), but in my use-case, the conditions for when the given nullable column would be NULL are well-known so it's not awful. 

Do you mean that passing knex.raw into the mikro-orm query builder could suffice? I'm trying to get an example of that working, but don't want to go too far down that road if that's not what you mean,

**Reply by B4nan** [reply]
Thanks, will try to look into that soon (but I got quite a busy week now).

> We cannot express the partial indexes here 

What's wrong with the index expressions you have in the comment there? That should work fine, if you need two, just use the decorator twice.

**Reply by chrisbenincasa** [reply]
I can make a separate test showing what is wrong, but the generated query seems to ignore the expression:

when I define the index like this:
```
@Unique({
  name: "a_null",
  properties: ["z", "y"],
  expression: "create unique index a_null (`z`, `y`) on a where `x` is NULL;",
})
```

I see the generated SQL:
```
console.log
    [query] create unique index `a_null` on `a` (`z`, `y`); [took 1 ms, 0 results]

      at DefaultLogger.log (node_modules/@mikro-orm/core/logging/DefaultLogger.js:38:14)
```

which seems to ignore the expression altogether. 

Dropping the properties field yields (testing this because I thought that maybe it would take the `expression` field as-is and move forward...):

```
console.log
    [query] create unique index `a_null` on `a` (); [took 0 ms]

      at DefaultLogger.log (node_modules/@mikro-orm/core/logging/DefaultLogger.js:38:14)
```

**EDIT**
Digging a little further, it seems like `expression` is ignored when using `Unique` but not `Index`

This works:
```typescript
@Index({
  name: "a_null",
  properties: ["z", "y"],
  expression: "create unique index a_null on `a` (`z`, `y`) where `x` is NULL",
})
@Index({
  name: "a_non_null",
  properties: ["z", "y"],
  expression:
    "create unique index a_non_null on `a` (`z`, `y`, `x`) where `x` is NOT NULL",
})
```

**Reply by B4nan** [reply]
This should fix the unique expressions 96dff537486359d968bb4da5e0e8d73ccdd89ad9

**Reply by B4nan** [reply]
I started refactoring the internals to be able to use `raw()` helper similarly to how knex supports it:

```ts
await em.upsert(
  A,
  { y: 'y1', z: 'z1' },
  {
    onConflictFields: sql`(z, y) where x is null`,
    onConflictAction: 'ignore',
    onConflictMergeFields: ['z'],
  },
);
```

It will take some time to finish, but hopefully will be in the next release (which will be v6.3, probably in a week or two).

**Reply by B4nan** [reply]
Here you go #5691

**Reply by chrisbenincasa** [reply]
This rocks! Thank you. I will check out this weekend. I have the workaround with just knex working but this will be a huge improvement.
